src/glview/preview/OpenCSGRenderer.cc

```diff
@@ -165,50 +165,25 @@ void OpenCSGRenderer::createCSGProducts(const CSGProducts& products, const Rende
                              all_vbos_[vbo_index++]);
+    VertexStateManager vsm(&(*vertex_states), &vertex_array);  // TODO Might be considered bad practice. But with smart pointers, then VSM could de-allocate these stack objects.
     vertex_array.addSurfaceData();
     vertex_array.writeSurface();
     add_shader_data(vertex_array);
 
-    if (Feature::ExperimentalVxORenderersDirect.is_enabled() || Feature::ExperimentalVxORenderersPrealloc.is_enabled()) {
-      size_t vertices_size = 0, elements_size = 0;
-      for (const auto& csgobj : product.intersections) {
-        if (csgobj.leaf->geom) {
-          vertices_size += getSurfaceBufferSize(csgobj, highlight_mode, background_mode, OpenSCADOperator::INTERSECTION);
-        }
-      }
-      for (const auto& csgobj : product.subtractions) {
-        if (csgobj.leaf->geom) {
-          vertices_size += getSurfaceBufferSize(csgobj, highlight_mode, background_mode, OpenSCADOperator::DIFFERENCE);
-        }
-      }
-
-      if (Feature::ExperimentalVxORenderersIndexing.is_enabled()) {
-        vertex_array.elementsVBO() = all_vbos_[vbo_index++];
-        if (vertices_size <= 0xff) {
-          vertex_array.addElementsData(std::make_shared<AttributeData<GLubyte, 1, GL_UNSIGNED_BYTE>>());
-        } else if (vertices_size <= 0xffff) {
-          vertex_array.addElementsData(std::make_shared<AttributeData<GLushort, 1, GL_UNSIGNED_SHORT>>());
-        } else {
-          vertex_array.addElementsData(std::make_shared<AttributeData<GLuint, 1, GL_UNSIGNED_INT>>());
-        }
-        elements_size = vertices_size * vertex_array.elements().stride();
-        vertex_array.elementsSize(elements_size);
+    size_t vertices_size = 0;
+    for (const auto& csgobj : product.intersections) {
+      if (csgobj.leaf->geom) {
+        vertices_size += getSurfaceBufferSize(csgobj, highlight_mode, background_mode, OpenSCADOperator::INTERSECTION);
       }
-
-      vertices_size *= vertex_array.stride();
-      vertex_array.verticesSize(vertices_size);
-
-      GL_CHECKD(glBindBuffer(GL_ARRAY_BUFFER, vertex_array.verticesVBO()));
-      GL_CHECKD(glBufferData(GL_ARRAY_BUFFER, vertices_size, nullptr, GL_STATIC_DRAW));
-      if (Feature::ExperimentalVxORenderersIndexing.is_enabled()) {
-        GL_CHECKD(glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, vertex_array.elementsVBO()));
-        GL_CHECKD(glBufferData(GL_ELEMENT_ARRAY_BUFFER, elements_size, nullptr, GL_STATIC_DRAW));
+    }
+    for (const auto& csgobj : product.subtractions) {
+      if (csgobj.leaf->geom) {
+        vertices_size += getSurfaceBufferSize(csgobj, highlight_mode, background_mode, OpenSCADOperator::DIFFERENCE);
       }
-    } else if (Feature::ExperimentalVxORenderersIndexing.is_enabled()) {
-      vertex_array.elementsVBO() = all_vbos_[vbo_index++];
-      vertex_array.addElementsData(std::make_shared<AttributeData<GLuint, 1, GL_UNSIGNED_INT>>());
     }
 
+    vsm.initializeSize(vertices_size, all_vbos_, vbo_index);
```

Comment:
To make this code easier to read, would it make sense to instead make this call `vsm.initializeSize(vertices_size, all_vbos_[vbo_index++])`. Letting the caller manage this feels a bit cleaner.

Reply:
`vbo_index` is incremented multiple times ([here](https://github.com/openscad/openscad/pull/4675/files#diff-c7bd8e779ecb300b4770befff604ab40bd6912b4622040d2fc4734d33190c12dR18) and [here](https://github.com/openscad/openscad/pull/4675/files#diff-c7bd8e779ecb300b4770befff604ab40bd6912b4622040d2fc4734d33190c12dR45)). In theory, I could split `initializeSize` up into multiple functions to avoid this. But what I'm trying to do now is find large swatches of shared code between renderers (or slightly different code that can be consolidated with the overload-and-helper-method approach that I've proposed here).

Reply:
I see now that the increment is also conditional on enabled features. Let's keep it like this and we can revisit API design when the bigger picture starts to become more clear.